proposals/Exceptions.md

```diff
@@ -143,42 +93,85 @@ Exception indices are used by:
-
-### Try and catch blocks
+### Try-catch blocks
 
 A _try block_ defines a list of instructions that may need to process exceptions
 and/or clean up state when an exception is thrown. Like other higher-level
 constructs, a try block begins with a `try` instruction, and ends with an `end`
-instruction. That is, a try block is sequence of instructions having the
+instruction. That is, a try-catch block is sequence of instructions having the
 following form:
 
 ```
 try blocktype
   instruction*
-catch
+catch i
+  instruction*
+catch j
+  instruction*
+...
+catch_all
   instruction*
 end
 ```
 
-A try block ends with a `catch block` that is defined by the list of
-instructions after the `catch` instruction.
+A try-catch block contains zero or more `catch` blocks and zero or one
+`catch_all` block. All `catch` blocks should precede the `catch_all` block, if
+any. The `catch`/`catch_all` instructions (within the try construct) are called
+the _catching_ instructions. There should be at least `catch` or `catch_all`
+block within a try-catch block.
+
+The _body_ of the try block is the list of instructions before the first
+catching instruction. The _body_ of each catch block is the sequence of
+instructions following the corresponding catching instruction before the next
+catching instruction (or the `end` instruction if it is the last catching
+block).
+
+The `catch` instruction has an exception tag associated with it. The tag
+identifies what exceptions it can catch. That is, any exception created with the
+corresponding exception tag. Catch blocks that begin with a `catch` instruction
+are considered _tagged_ catch blocks.
+
+The last catching instruction of a try-catch block can be the `catch_all`
+instruction. If it begins with the `catch_all` instruction, it defines the
+_default_ catch block. The default catch block has no exception type, and is
+used to catch all exceptions not caught by any of the tagged catch blocks. The
+term 'catch block' refers to both `catch` and `catch_all` blocks.
 
 Try blocks, like control-flow blocks, have a _block type_. The block type of a
 try block defines the values yielded by evaluating the try block when either no
 exception is thrown, or the exception is successfully caught by the catch block.
 Because `try` and `end` instructions define a control-flow block, they can be
 targets for branches (`br` and `br_if`) as well.
 
-In the initial implementation, try blocks may only yield 0 or 1 values.
+### Try-unwind blocks
+
+Try blocks can also be used with the `unwind` instruction. A try-unwind block
+contains an `unwind` block with the following form:
+
+```
+try blocktype
+  instruction*
+unwind
+  instruction*
+end
+```
+
+The `unwind` block is meant to contain cleanup instructions, such as
+destructors, in case any instruction in the corresponding try block throws.
+Currently the `unwind` instruction has the same semantics as the `catch_all`
```

Comment:
I'm surprised and confused with these semantics of `unwind`.

If `try ... unwind ... end` is semantically equivalent to `try ... catch_all ... rethrow end`, then I can't see why we are adding this. Can't we just postpone this to a future two-phase unwinding proposal? In fact, if the `unwind` semantics will change later, couldn't that create backwards compatibility problems?

Until this document, I had understood `unwind` more as a primitive related to common-lisp's `unwind-protect`, which btw [seems to be related to]( https://stackoverflow.com/a/12999898 ) Java's and Python's `finally` instead of `catch_all`.

Reply:
Also it seems that `unwind` could spark a very long discussion, for example when common-lisp was being specified, `unwind-protect` discussions spanned over years ([see this link](http://www.lispworks.com/documentation/HyperSpec/Issues/iss151_w.htm), at the bottom of the page). 

If we can avoid adding this instruction now, we could have these discussions in a different proposal.

Reply:
@ioannad It's good to point out that there are good reasons to be wary of adding control constructs. Fortunately, though, in the case of `unwind`, the control construct has been tried out in a wide variety of languages (with `finally` being a close relative) and its semantics are fairly well understood. It seems to have been found to be particularly useful in multi-language settings for reasons akin to why it was added here to make modules forwards compatible with two-phase exception handling. And nowadays its semantics are fairly well understood, at least at the level that WebAssembly operates at. (This wasn't the case back when Common Lisp was being specified, and some of the complications in the referenced discussion have to do with higher-level considerations.)

I cannot say the same for `catch_all` or `rethrow` though, so rather than removing `unwind` due to the above equivalence, my suggestion would be to remove `catch_all` instead.

Reply:
@ioannad They have the same semantics, and they are planned to have different semantics in two-phase. In single-phase unwinding, both `catch_all`'s and `unwind`'s bodies run as we search up and unwind the stack. In two-phase, `catch` bodies run after we finish all two-phase unwinding process, while `unwind` bodies run as cleanups in the second phase, and their `end` instruction act as "resuming the second phase" instruction.

I think what you said also makes sense that if two are equivalent, why adding them now? I don't have a strong opinion either way myself. It was taken from @RossTate's suggestion, and I agreed with that for the reason that it'd be good if we make the current proposal and the follow-on future two-phase proposal as similar as possible.

@RossTate As I said in the other issue, `catch_all` has its uses. If we remove that, `unwind` cannot replace it when two-phase arrives. These days, I am almost getting an impression that you are trying to remove every single instruction that you are not gonna use for your future proposals..

Reply:
> though, in the case of `unwind`, the control construct has been tried out in a wide variety of languages

@RossTate, can you point us to any such language? Constructs like `finally` or `unwind-protect` seem much more common-place and semantically well-understood.

Reply:
What is called `unwind` here is called `fault` in the .NET CLI. Like `finally`, `fault`/`unwind` is executed whenever the stack (frame) is unwound (as opposed to just basic cleanup like deallocated). The only difference between `finally` and `fault` is that the former is also executed by local control, e.g. `br` and `return`. We need to make the distinction in this proposal because the semantics of C++ destructors is different for the two kinds of paths.

Reply:
AFAIK, `fault` is not used by C#. Do we have any information where it is used? Managed C++?

Wouldn't the distinction needed by C++ be aided better by a `finally` that passes in a Boolean to indicate whether it's regular or exceptional finalisation?

Reply:
From what I understand, you would have to make significant changes to LLVM to get that to happen, including changing how it represents CFGs and exception handling. My take is that `finally` is a surface-level construct, and that `unwind` is part of its lower-level counterpart, with code duplication or #124 providing ways to support the rest of its lower-level counterpart.

Reply:
Well, the LLVM backend gets by without any use of finally/unwind right now, so, shrug? And of course it is trivial to express fault a.k.a. unwind with such a flagged finally if needs be.

I'd still be interested in knowing which .NET producers actually use fault.

Reply:
> Well, the LLVM backend gets by without any use of finally/unwind right now, so, shrug?

Every function call in LLVM has an argument specifying how to unwind it.

> I'd still be interested in knowing which .NET producers actually use fault.

It's how .NET addresses the problem above that C++ destructors have different semantics on the two kinds of exit paths.

Reply:
About LLVM: It compiles destructors for the normal path and the exception path separately, because the exception path contains some different logic, such as, if another exception arises within there, it does not throw but calls `std::terminate` or something. This is C++, but it also compiles `finally` away for some languages that have `finally`, such as Objective-C.

In the CFG level, LLVM has an `invoke` instruction, which is a variant of `call` instruction. `invoke` takes two successors: a normal destination, where we should go when the call returns normally, and unwind destination, where we should unwind to when the call throws. Our wasm backend lowers this to `try`-`catch`.

So my point is, if we introduce `finally` in replacement for `unwind`, at least for LLVM it is not very usable for us. And I am not 100% sure if I understood your fragmented stack argument, but I think `finally` has all the same problem anyway, no?

We can still use `catch` for cleanup code as we have so far, but for 2PEH we need to distinguish user catch handler from cleanup code. That doesn't have to be `unwind`, but what I'm saying is we need a way.

Reply:
Yes, I can see that it's not useful with the specific choices LLVM made. But other compilers could benefit. If there is pressure to add finally, then we shouldn't end up having both unwind and finally, thus the idea of a simple generalisation that expresses both.

> And I am not 100% sure if I understood your fragmented stack argument, but I think finally has all the same problem anyway, no?

That is unrelated. That problem problem occurs with the search phase, not the unwind phase. The latter runs on the top of the stack, since all other stack frames are already unwound at that point.